src/src_cryptography/src_libgcrypt/cryptography_interface_libgcrypt.template.c

```diff
@@ -728,10 +732,17 @@ static int32_t cryptography_validate_authentication(uint8_t* data_out, size_t le
     if ((gcry_error & GPG_ERR_CODE_MASK) != GPG_ERR_NO_ERROR)
     {
         printf(KRED "ERROR: gcry_mac_read error code %d\n" RESET, gcry_error & GPG_ERR_CODE_MASK);
-        printf(KRED "Failure: %s/%s\n", gcry_strsource(gcry_error), gcry_strerror(gcry_error));
+        printf(KRED "Failure: %s/%s\n" RESET, gcry_strsource(gcry_error), gcry_strerror(gcry_error));
+        gcry_mac_close(tmp_mac_hd);
         status = CRYPTO_LIB_ERR_MAC_RETRIEVAL_ERROR;
         return status;
     }
+    // Now that MAC has been verified, check IV & ARSN if applicable
+    if (crypto_config->ignore_anti_replay == TC_IGNORE_ANTI_REPLAY_FALSE)
+    {
+        status = Crypto_Check_Anti_Replay(sa_ptr, arsn, iv);
```

Comment:
The anti-replay counter check is not specific to a cryptography interface... This check belongs in crypto_tc.c, not in the libgcrypt/kmc crypto template files.

Reply:
Or is there some reason that I'm missing about why this needs to live here?

Reply:
This is a good point, looking into it.

Reply:
Ok, so I recall now - the reason I did this here is that we cannot validate the ARSN until we know the MAC (if present) is valid - which requires several libgcrypt calls to open/key/validate. Rather than having half of the ARSN checks in crypto_tc.c, and the other half in the decryption crypto_ifs, I opted to place them all the same way. But, I'm open to discussion and more thoughts on this

Reply:
I'm tracking better what you're saying now.... so it's fine if we decrypt the text, and then just discard it if the ARSN/IV turned out be be invalid

Reply:
I would think so... for AEAD algorithms, we don't have a choice... for non-AEAD algorithms, we can still have the logic between the authenticate and decrypt calls if we wanted.